### PR TITLE
Setup a changelog that pulls from PR description

### PR DESCRIPTION
**Pull request by mwaskom**
## Describe your changes

This sets up the basic infrastructure to start recording changes in the client on a per-version basis. Once this is merged, the PR template will include a "Changelog" section. For user-facing changes (new features, bug fixes, deprecations, etc.), we can add a short description in this section. Once the PR is merged, the action that triggers on a push to main will fetch the PR description and insert it into `CHANGELOG.md`. The section is optional, and the action should abort if you delete the "Changelog" section or simply leave it empty.

The changelog is currently organized with h3 headers (containing the modal version and date) for each micro release. Because the action simply prepends new information, we can revise the existing content or add new content manually if necessary (provided that we are careful to preserve the token that marks where new content should be inserted!)

Please review carefully: I developed this in a test repo and everything looked good, but we won't be able to exercise it in the modal-client repo until it's actually merged, and it could break the automatic release process if it fails on an edge case.

Resolves MOD-2054

### CODE REVIEW

**Comment by ekzhang** [comment]
We can add the changelog via entering into a box if we use a YAML issue template https://github.com/google/gvisor/blob/master/.github/ISSUE_TEMPLATE/bug_report.yml

File: tasks.py [hunk 1]
```diff
@@ -15,6 +16,7 @@
 from pathlib import Path
 from typing import Optional
 
+import requests
```
Comment (ekzhang) [inline tasks.py:19]: Is requests installed?

File: tasks.py [hunk 2]
```diff
@@ -155,3 +157,60 @@ def type_stubs(ctx):
...
+        break_idx = m.span()[1]
+        header = content[:break_idx]
+        previous_changelog = content[break_idx:]
+    else:
+        print("Aborting: Could not find token in existing changelog to mark insertion spot")
+
+    # Build the new changelog and write it out
+    from modal_version import __version__
+
+    date = datetime.now().strftime("%Y-%m-%d")
```
Comment (ekzhang) [inline tasks.py:212]: Maybe `%B %-d, %Y` for

## 0.55.5069 (January 10, 2024)

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+This changelog documents user-facing changes (features, fixes, and deprecations) in `modal`.
```
Comment (ekzhang) [inline CHANGELOG.md:3]: > in _the Modal client library_.

How do you feel about the copy from https://github.com/modal-labs/modal-client/blob/387410ac6c5b860cde32f3cd8f96a9a16d77646d/CHANGELOG.md — users often ask us about our version support policy so it would be nice to include a sentence?

**Review by ekzhang — APPROVED** [review]
nice!

File: tasks.py (hunk 2 above)
Comment (erikbern) [reply tasks.py:212]: more of an iso 8601 fan

**Comment by erikbern** [comment]
Looks good! What's the flow here, is the intent that we manually edit these changelogs as well? Or just rely on fully automated ones

File: CHANGELOG.md (hunk 3 above)
Comment (mwaskom) [reply CHANGELOG.md:3]: Yeah good thinking! Where are you pulling the **Historical support matrix** from here? I think that @howard36 has been doing some work on defining support windows as well, I wonder how that intersects with this.

File: tasks.py (hunk 1 above)
Comment (mwaskom) [reply tasks.py:19]: I think so, on the basis of the `setup-cached-python` step that runs before this. It worked when I copy-pasted that action source into my test repo. That said, it looks like requests is just a transitive dependency and not in on `requirements` file. Let me know if you think we need to add it there or if my understanding of the setup flow is otherwise off.

File: tasks.py [hunk 4]
```diff
@@ -155,3 +157,60 @@
...
+    url = f"https://api.github.com/repos/modal-labs/modal-client/pulls/{pull_number}"
+    headers = {"Authorization": f"Bearer {os.environ['GITHUB_TOKEN']}", "Accept": "application/vnd.github.v3+json"}
+    response = requests.get(url, headers=headers).json()
+    pr_description = response.get("body")
+    if pr_description is None:
+        print("Aborting: No PR description in response from GitHub API")
+        return
+
+    # Parse the PR description to get a changelog update
+    comment_pattern = r"<!--.+?-->"
```
[the remark this answers is not in the document]
Comment (mwaskom) [reply tasks.py:183]: I think this is fine as it's just stripping out the instructions we add to the PR template but I guess an edge case would be if someone expands those in the future using a multi-line comment.

File: tasks.py (hunk 2 above)
Comment (mwaskom) [reply tasks.py:212]: 🧑‍⚖️

**Comment by mwaskom** [comment]
> Looks good! What's the flow here, is the intent that we manually edit these changelogs as well? Or just rely on fully automated ones

I think we'll want to see how things go maybe. In theory I don't think we'd need to make manual edits, but a virtue of this approach is that we _can_ if we need to (i.e. it doesn't get auto-generated top-to-bottom every release). One use-case for manual editing that I can imagine is writing up some notes for minor (0.YY) releases to aggregate the key new features.

File: tasks.py (hunk 1 above)
Comment (mwaskom) [reply tasks.py:19]: That said I'm not totally sure how to interpret the failing mypy step, which pertains to `requests`...

Comment (mwaskom) [reply tasks.py:19]: Added `requests` and `types-requests` to the dev requirements.

File: CHANGELOG.md (hunk 3 above)
Comment (mwaskom) [reply CHANGELOG.md]: Copied some of your header text over!

We might need to think a little more about the best way to programmatically sync the deprecation schedule that @howard36 implemented with the changelog, since it currently lives as a config in the `modal` repo. I've left that out for now. (I guess there's also a subtle distinction between "when will functionality be removed from the client" and "when will the server no longer support older clients").

**Comment by ekzhang** [comment]
Re: syncing the deprecation schedule, it's probably fine to just manually update it. We only make changes every few months, and changing more often than that would confuse users!
